File: app/src/main/java/com/dumuzeyn/mp3player/LibraryMaintenanceController.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import android.os.Handler;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.atomic.AtomicBoolean;

/** Reconciles persisted tracks and repairs old metadata entirely off the UI thread. */
final class LibraryMaintenanceController implements AutoCloseable {
    static final int METADATA_REVISION = 1;

    interface Callback {
        void finished(List<Track> refreshed, List<Track> unavailable);
    }

    interface UnavailableCallback {
        void finished(List<Track> unavailable);
    }

    private Context context;
    private final Handler mainHandler;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final AtomicBoolean started = new AtomicBoolean();
    private volatile boolean closed;

    LibraryMaintenanceController(Context context, Handler mainHandler) {
        this.context = context;
        this.mainHandler = mainHandler;
    }

    void run(List<Track> source, Callback callback) {
        if (closed || !started.compareAndSet(false, true)) {
            return;
        }
        promoteApplicationContext();
        ArrayList<Track> snapshot = new ArrayList<>(source);
        try {
            executor.execute(() -> maintain(snapshot, callback));
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    void inspectUnavailable(List<Track> source, UnavailableCallback callback) {
        promoteApplicationContext();
        ArrayList<Track> snapshot = new ArrayList<>(source);
        try {
            executor.execute(() -> {
                ArrayList<Track> unavailable = unavailable(snapshot);
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.finished(unavailable);
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    void removeUnavailable(List<Track> source, Runnable callback) {
        promoteApplicationContext();
        ArrayList<Track> unavailable = new ArrayList<>(source);
        try {
            executor.execute(() -> {
                TrackStore.applyMaintenance(context, new ArrayList<>(), new HashSet<>(),
                        unavailable, METADATA_REVISION);
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.run();
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    @Override
    public void close() {
        closed = true;
        executor.shutdownNow();
    }

    private void maintain(ArrayList<Track> tracks, Callback callback) {
        ArrayList<Track> refreshed = new ArrayList<>();
        ArrayList<Track> unavailable = new ArrayList<>();
        HashSet<String> checked = new HashSet<>();
        Map<String, Track> candidates = candidatesById();
        for (Track track : tracks) {
            if (closed) {
                return;
            }
            LibraryFileAccessManager.AccessState access =
                    LibraryFileAccessManager.accessState(context, track);
            if (access == LibraryFileAccessManager.AccessState.UNAVAILABLE) {
                unavailable.add(track);
                continue;
            }
            if (access != LibraryFileAccessManager.AccessState.AVAILABLE
                    || !candidates.containsKey(track.trackId)) {
                continue;
            }
            checked.add(track.trackId);
            if (!needsRefresh(track)) {
                continue;
            }
            Track updated = TrackStore.refreshMetadata(context, track);
            if (metadataChanged(track, updated)) {
                refreshed.add(updated);
            }
        }
        if (closed) {
            return;
        }
        TrackStore.applyMaintenance(context, refreshed, checked, unavailable,
                METADATA_REVISION);
        mainHandler.post(() -> {
            if (!closed) {
                callback.finished(refreshed, unavailable);
            }
        });
    }

    private ArrayList<Track> unavailable(List<Track> tracks) {
        ArrayList<Track> result = new ArrayList<>();
        for (Track track : tracks) {
            if (closed) {
                break;
            }
            if (LibraryFileAccessManager.accessState(context, track)
                    == LibraryFileAccessManager.AccessState.UNAVAILABLE) {
                result.add(track);
            }
        }
        return result;
    }

    private Map<String, Track> candidatesById() {
        HashMap<String, Track> result = new HashMap<>();
        for (Track track : TrackStore.loadMetadataRefreshCandidates(
                context, METADATA_REVISION)) {
            result.put(track.trackId, track);
        }
        return result;
    }

    private void promoteApplicationContext() {
        Context application = context.getApplicationContext();
        if (application != null) {
            context = application;
        }
    }

    private static boolean needsRefresh(Track track) {
        return track.durationMs <= 0
                || isMissing(track.artist, "Unknown artist")
                || isMissing(track.album, "Unknown album")
                || GenreNormalizer.isUnknown(track.genre);
    }

    private static boolean isMissing(String value, String placeholder) {
        return value == null || value.trim().isEmpty() || placeholder.equalsIgnoreCase(value.trim())
                || "<unknown>".equalsIgnoreCase(value.trim());
    }

    private static boolean metadataChanged(Track before, Track after) {
        return before.durationMs != after.durationMs
                || !same(before.artist, after.artist)
                || !same(before.album, after.album)
                || !same(before.albumArtist, after.albumArtist)
                || !same(before.genre, after.genre)
                || before.year != after.year
                || before.trackNumber != after.trackNumber
                || before.discNumber != after.discNumber;
    }

    private static boolean same(String left, String right) {
        return left == null ? right == null : left.equals(right);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/LibraryMaintenanceController.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.os.Handler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicBoolean

/** Reconciles persisted tracks and repairs old metadata entirely off the UI thread. */
class LibraryMaintenanceController(
    private val ownerContext: Context,
    private val mainHandler: Handler,
) : AutoCloseable {
    fun interface Callback {
        fun finished(refreshed: List<Track>, unavailable: List<Track>)
    }

    fun interface UnavailableCallback {
        fun finished(unavailable: List<Track>)
    }

    private val context: Context by lazy(LazyThreadSafetyMode.NONE) {
        ownerContext.applicationContext ?: ownerContext
    }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val started = AtomicBoolean()
    @Volatile private var closed = false

    fun run(source: List<Track>, callback: Callback) {
        if (closed || !started.compareAndSet(false, true)) return
        val snapshot = ArrayList(source)
        scope.launch { maintain(snapshot, callback) }
    }

    fun inspectUnavailable(source: List<Track>, callback: UnavailableCallback) {
        val snapshot = ArrayList(source)
        scope.launch {
            val unavailable = unavailable(snapshot)
            mainHandler.post {
                if (!closed) callback.finished(unavailable)
            }
        }
    }

    fun removeUnavailable(source: List<Track>, callback: Runnable) {
        val unavailable = ArrayList(source)
        scope.launch {
            TrackStore.applyMaintenance(
                context,
                ArrayList(),
                HashSet(),
                unavailable,
                METADATA_REVISION,
            )
            mainHandler.post {
                if (!closed) callback.run()
            }
        }
    }

    override fun close() {
        closed = true
        scope.cancel()
    }

    private fun maintain(tracks: ArrayList<Track>, callback: Callback) {
        val refreshed = ArrayList<Track>()
        val unavailable = ArrayList<Track>()
        val checked = HashSet<String>()
        val candidates = candidatesById()
        for (track in tracks) {
            if (closed) return
            when (LibraryFileAccessManager.accessState(context, track)) {
                LibraryFileAccessManager.AccessState.UNAVAILABLE -> {
                    unavailable += track
                    continue
                }
                LibraryFileAccessManager.AccessState.DEFERRED -> continue
                LibraryFileAccessManager.AccessState.AVAILABLE -> Unit
            }
            if (track.trackId !in candidates) continue
            checked += track.trackId
            if (!needsRefresh(track)) continue
            val updated = TrackStore.refreshMetadata(context, track)
            if (metadataChanged(track, updated)) refreshed += updated
        }
        if (closed) return
        TrackStore.applyMaintenance(context, refreshed, checked, unavailable, METADATA_REVISION)
        mainHandler.post {
            if (!closed) callback.finished(refreshed, unavailable)
        }
    }

    private fun unavailable(tracks: List<Track>): ArrayList<Track> {
        val result = ArrayList<Track>()
        for (track in tracks) {
            if (closed) break
            if (
                LibraryFileAccessManager.accessState(context, track) ==
                LibraryFileAccessManager.AccessState.UNAVAILABLE
            ) {
                result += track
            }
        }
        return result
    }

    private fun candidatesById(): Map<String, Track> =
        TrackStore.loadMetadataRefreshCandidates(context, METADATA_REVISION)
            .associateBy(Track::trackId)

    private fun needsRefresh(track: Track): Boolean =
        track.durationMs <= 0 ||
            isMissing(track.artist, "Unknown artist") ||
            isMissing(track.album, "Unknown album") ||
            GenreNormalizer.isUnknown(track.genre)

    private fun isMissing(value: String?, placeholder: String): Boolean {
        val cleaned = value?.trim().orEmpty()
        return cleaned.isEmpty() ||
            cleaned.equals(placeholder, ignoreCase = true) ||
            cleaned.equals("<unknown>", ignoreCase = true)
    }

    private fun metadataChanged(before: Track, after: Track): Boolean =
        before.durationMs != after.durationMs ||
            before.artist != after.artist ||
            before.album != after.album ||
            before.albumArtist != after.albumArtist ||
            before.genre != after.genre ||
            before.year != after.year ||
            before.trackNumber != after.trackNumber ||
            before.discNumber != after.discNumber

    companion object {
        const val METADATA_REVISION = 1
    }
}
